Several UX tweaks for riot

Yurlungur opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "Add AMR unit test for cell centered fields.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

We are ramping up to make riot more broadly available for LANL users internally and I'm going through some wishlist items we want to expose in parthenon for riot for that. I think these features are broadly useful/uncontroversial:
- I added a per-package PostInitialize function.
- I added a "magnitude" refinement criteria so you can refine or derefine on the magnitude of some field. Note that for sparse fields, you will need to specify the sparse index which is maybe a little annoying.
- I cleaned up the refinement criteria logic so refinement criteria no longer need to be sequential, similar to how outputs work.
- I added a `pyroject.toml` file for modern python package installation. (I didn't remove the old setup.py. I did change it a little.) 
- I also added the user-accessible scripts to the scripts list for installation so they get installed into somewhere accessible by PATH. And I gave them executable permissions.
- I added an optional log scale to `movie2d.py`
- I added two python scripts designed to work with docstrings:
  - `mergecsv.py` can combine param csv files output by parthenon and print a new csv file to stdout that has sorted them and removed duplicates.
  - `prettyparams.py` can read a params csv output by parthenon, either as a file or from stdin. It can then either print out a pretty ASCII table, like this one:

```
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| block                        | parameters                  | type   | default            | description               |
+==============================+=============================+========+====================+===========================+
| diffusion                    | boundary_prolongation       | string | Linear             |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | cfl                         | Real   | 1                  |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | constant_coefficient        | bool   | 1                  |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | diagonal_alpha              | Real   | 1                  |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | dt                          | Real   | 1                  |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | flux_correct                | bool   | 0                  |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | set_flux_boundary           | bool   | 0                  |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | solver                      | string | MG                 |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | t0                          | Real   | 0.001              |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | x0                          | Real   | 0                  |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
| diffusion                    | y0                          | Real   | 0                  |                           |
+------------------------------+-----------------------------+--------+--------------------+---------------------------+
```

or it can open into a CLI-based interactive menu that can be searched and browsed, and where blocks can be hidden or shown. Here's a screenshot:
<img width="1201" height="687" alt="image" src="https://github.com/user-attachments/assets/61804a25-d581-4842-bb82-9b882a227021" />

Note this is based on @acreyes MR #1307 as I wanted this all to work if you installed `parthenon_tools`.

Also note I relied heavily on Gen AI for the two new python scripts, as the robots do a pretty good job with menial/boilerplate tasks like this. I put a disclaimer in file.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Code passes cpplint
- [x] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] Change is breaking (API, behavior, ...)
  - [ ] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [ ] PR is marked as breaking
  - [ ] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] Docs build
- [x] (@lanl.gov employees) Update copyright on changed files

acreyes committed:
invert -> tol < err

acreyes committed:
fix imports

acreyes committed:
relative imports instead

acreyes committed:
install parthenon_tools in CI

jonahm-LANL committed:
movie2d features

Yurlungur committed:
Merge branch 'develop' into acreyes/parthenon-tools

jonahm-LANL committed:
pip install --user

jonahm-LANL committed:
CI short too

jonahm-LANL committed:
I know what I'm doing debian. --break-system-packages it is

jonahm-LANL committed:
changelog

acreyes committed:
absolute imports

jonahm-LANL committed:
merge csv and pretty params

jonahm-LANL committed:
pretty params

jonahm-LANL committed:
make them executable

jonahm-LANL committed:
add a pyproject.toml to avoid setup.py but keep the old one for backwards compatibility

jonahm-LANL committed:
remove parthenon tools install from ci extended

jonahm-LANL committed:
Merge branch 'acreyes/parthenon-tools' into jmm/ux-tweaks

jonahm-LANL committed:
amr criteria and post-initialization per package

jonahm-LANL committed:
docs

jonahm-LANL committed:
formatting and linting

jonahm-LANL committed:
changelog

jonahm-LANL committed:
add a version to make pyproject.toml happy

adamdempsey90 commented on src/amr_criteria/refinement_package.cpp:
```diff
@@ -203,6 +203,47 @@ void SecondDerivative(const AMRBounds &bnds, MeshData<Real> *md, const std::stri
+              r = std::max(r, val);
+            },
+            Kokkos::Max<Real>(maxval));
+        auto tags_access = scatter_tags.access();
+        auto flag = AmrTag::same;
+        if (maxval > sign * refine_criteria && pack.GetLevel(b, 0, 0, 0) < max_level)
+          flag = AmrTag::refine;
+        if (maxval < sign * derefine_criteria) flag = AmrTag::derefine;
+        // ScatterMax view will use an atomic_max to prevent race condition across k,j
+        // indices
+        tags_access(b).update(flag);
```
I hate the scatterview interface so much. The comment is incorrect for an openmp & serial backends, technically.

Yurlungur replied:
Oof ok. I'll just remove the comment. I blindly copy-pasted it from the other criteria.

acreyes replied:
that one is my bad, I think I wrote the initial comment this is copied from.

Yurlungur replied:
comment removed

pdmullen commented on src/mesh/mesh.cpp:
```diff
@@ -818,6 +818,29 @@ void Mesh::Initialize(bool init_problem, ParameterInput *pin, ApplicationInput *
         }
       }
 
+      // Call per-package initialization
+      for (const auto &[name, pkg] : packages.AllPackages()) {
+        PARTHENON_REQUIRE_THROWS(
+            !(pkg->PostInitializationMesh != nullptr &&
+              (pkg->PostInitializationBlock != nullptr)),
+            "Mesh and MeshBlock PostInitializations are defined for package " + name +
+                ". Please use only one.");
+
+        // first on the mesh...
+        if (pkg->PostInitializationMesh != nullptr) {
+          for (auto &partition : GetDefaultBlockPartitions(GridIdentifier::leaf())) {
+            auto &md = mesh_data.Add("base", partition);
+            pkg->PostInitializationMesh(this, pin, md.get());
+          }
+        }
+        // and then per block
+        if (pkg->PostInitializationBlock != nullptr) {
+          for (auto &pmb : block_list) {
+            pkg->PostInitializationBlock(pmb.get(), pin);
+          }
+        }
+      }
+
```
Can we switch to the original `PostInitialization` routines getting the final say (i.e., they come last in call chain)? My current useage of this utility might benefit from this.

Yurlungur replied:
The reason I have the original ones taking priority is I thought that the base one, which is always on, should be overwritten by the per-package ones... And it's also consistent with, e.g., how our boundary conditions work. What did you have in mind for the other ordering?

Yurlungur replied:
Discussed in person and I think we're both happy keeping this ordering for now

pdmullen commented on src/amr_criteria/amr_criteria.hpp:
```diff
@@ -57,6 +58,24 @@ struct AMRSecondDerivative : public AMRCriteria {
   void operator()(MeshData<Real> *md, ParArray1D<AmrTag> &delta_level) const override;
 };
 
+struct AMRMagnitude : public AMRCriteria {
```
A more general magnitude criteria might permit:

refine_ above: val1
derefine_below: val2

where intermediate val2 < field < val1 returns `AMRTag::same`

pdmullen replied:
ah my bad... this is actually what is currently implemented since you still check on `refine_tol` and `derefine_tol`

jonahm-LANL committed:
remove incorrect comment about scattermax

jonahm-LANL committed:
apparently they need mains

jonahm-LANL committed:
fix allowed values in docstrings

jonahm-LANL committed:
fix bug where allowed vals were not being output

Yurlungur commented:
Anyone have any idea why the diffusion test is timing out? @lroberts36 @pgrete ? Maybe a CI blip?

Yurlungur committed:
Merge branch 'develop' into jmm/ux-tweaks

jonahm-LANL committed:
Have ToSizeN return sopmething sensible in the case of no meshblocks

jonahm-LANL committed:
Merge branch 'jmm/ux-tweaks' into jmm/fix-block-less-than-ranks

jonahm-LANL committed:
Merge branch 'jmm/ux-tweaks' of github.com:parthenon-hpc-lab/parthenon into jmm/ux-tweaks

Yurlungur commented:
I snuck in a fix for issue #1301 also. @chadmeyer I tested in riot but you could also give it a shot

pgrete commented:
```
  Start 48: Check that partitioning a container works
48/58 Test #48: Check that partitioning a container works ..................................***Failed    0.35 sec
Filters: Check that partitioning a container works

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
unit_tests is a Catch v2.13.8 host application.
Run with -? for options

-------------------------------------------------------------------------------
Check that partitioning a container works
      Given: An attempt to partition into zero partitions
       Then: The partition attempt throws an error
-------------------------------------------------------------------------------
/__w/parthenon/parthenon/tst/unit/test_partitioning.cpp:92
...............................................................................

/__w/parthenon/parthenon/tst/unit/test_partitioning.cpp:93: FAILED:
  REQUIRE_THROWS_AS( parthenon::partition::ToSizeN(v, psize), std::runtime_error )
because no exception was thrown where one was expected:
```
seems legit, doesn't it?

Yurlungur commented:
> ```
> 
>   Start 48: Check that partitioning a container works
> 
> 48/58 Test #48: Check that partitioning a container works ..................................***Failed    0.35 sec
> 
> Filters: Check that partitioning a container works
> 
> 
> 
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> unit_tests is a Catch v2.13.8 host application.
> 
> Run with -? for options
> 
> 
> 
> -------------------------------------------------------------------------------
> 
> Check that partitioning a container works
> 
>       Given: An attempt to partition into zero partitions
> 
>        Then: The partition attempt throws an error
> 
> -------------------------------------------------------------------------------
> 
> /__w/parthenon/parthenon/tst/unit/test_partitioning.cpp:92
> 
> ...............................................................................
> 
> 
> 
> /__w/parthenon/parthenon/tst/unit/test_partitioning.cpp:93: FAILED:
> 
>   REQUIRE_THROWS_AS( parthenon::partition::ToSizeN(v, psize), std::runtime_error )
> 
> because no exception was thrown where one was expected:
> 
> ```
> 
> seems legit, doesn't it?

Ah, this is testing functionality I changed. The test should change actually. I changed this to allow  parthenon to start with fewer blocks than ranks, so long as after AMR there are sufficient blocks.

jonahm-LANL committed:
try this solution

Yurlungur committed:
Merge branch 'develop' into jmm/ux-tweaks

Yurlungur commented:
> > ```
> > 
> >   Start 48: Check that partitioning a container works
> > 
> > 48/58 Test #48: Check that partitioning a container works ..................................***Failed    0.35 sec
> > 
> > Filters: Check that partitioning a container works
> > 
> > 
> > 
> > ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> > 
> > unit_tests is a Catch v2.13.8 host application.
> > 
> > Run with -? for options
> > 
> > 
> > 
> > -------------------------------------------------------------------------------
> > 
> > Check that partitioning a container works
> > 
> >       Given: An attempt to partition into zero partitions
> > 
> >        Then: The partition attempt throws an error
> > 
> > -------------------------------------------------------------------------------
> > 
> > /__w/parthenon/parthenon/tst/unit/test_partitioning.cpp:92
> > 
> > ...............................................................................
> > 
> > 
> > 
> > /__w/parthenon/parthenon/tst/unit/test_partitioning.cpp:93: FAILED:
> > 
> >   REQUIRE_THROWS_AS( parthenon::partition::ToSizeN(v, psize), std::runtime_error )
> > 
> > because no exception was thrown where one was expected:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > seems legit, doesn't it?
> 
> Ah, this is testing functionality I changed. The test should change actually. I changed this to allow parthenon to start with fewer blocks than ranks, so long as after AMR there are sufficient blocks.

Ah actually no. I think there's a better solution in `mesh.cpp` Testing.

jonahm-LANL committed:
come on C++ do better

jonahm-LANL committed:
found the offending 0 and removed it

Yurlungur commented:
> ```
> 
>   Start 48: Check that partitioning a container works
> 
> 48/58 Test #48: Check that partitioning a container works ..................................***Failed    0.35 sec
> 
> Filters: Check that partitioning a container works
> 
> 
> 
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> unit_tests is a Catch v2.13.8 host application.
> 
> Run with -? for options
> 
> 
> 
> -------------------------------------------------------------------------------
> 
> Check that partitioning a container works
> 
>       Given: An attempt to partition into zero partitions
> 
>        Then: The partition attempt throws an error
> 
> -------------------------------------------------------------------------------
> 
> /__w/parthenon/parthenon/tst/unit/test_partitioning.cpp:92
> 
> ...............................................................................
> 
> 
> 
> /__w/parthenon/parthenon/tst/unit/test_partitioning.cpp:93: FAILED:
> 
>   REQUIRE_THROWS_AS( parthenon::partition::ToSizeN(v, psize), std::runtime_error )
> 
> because no exception was thrown where one was expected:
> 
> ```
> 
> seems legit, doesn't it?

Fixed. There was a cleaner solution than changing utils
